Automattic/wp-calypso
# Framework: Substitute crypto polyfill dependants with smaller stand-ins

A few of our dependencies, namely `component-uid` and `js-sha1`, themselves have dependencies on `crypto`:

https://github.com/emn178/js-sha1/blob/c724643/src/sha1.js#L52-L53
https://github.com/component/uid/blob/2d912b8/index.js#L45 (It appears they try to avoid the polyfill, but it _is_ included in a Webpack build)

`crypto` is a [Node built-in module](https://nodejs.org/api/crypto.html) which is polyfilled in the browser by [crypto-browserify](https://github.com/crypto-browserify/crypto-browserify). This polyfill is quite substantial in size.

The changes here seek to find suitable replacements for each: `uuid` and [`hash.js`](https://www.npmjs.com/package/hash.js) respectively, which are very reasonable in size.

_Aside: I first suspected `uuid` to also [require `crypto`](https://github.com/kelektiv/node-uuid/blob/72fbabbf00192817bdce7073e323361f5e7bdd67/lib/rng.js#L4), but it turns out they bundle a [browser-specific variant](https://github.com/kelektiv/node-uuid/blob/72fbabbf00192817bdce7073e323361f5e7bdd67/lib/rng-browser.js) without the dependency_.

The difference is pretty noticeable: 472kb smaller minified (112kb gzipped) in the main `build.js`.

**Before:**

```
wp-calypso (master) $ wc -c public/build.b1e9a12a27f881b1406e.js
 7190661 public/build.b1e9a12a27f881b1406e.js
wp-calypso (master) $ gzip-size public/build.b1e9a12a27f881b1406e.js --raw
1275631
```

**After:**

```
wp-calypso (master) $ wc -c public/build.10d51aa5c5ffef307acf.js
 6718680 public/build.10d51aa5c5ffef307acf.js
wp-calypso (master) $ gzip-size public/build.10d51aa5c5ffef307acf.js --raw
1163240
```

__Testing instructions:__

Verify against regressions for the two modules affected:

- Sync handler (used for caching response data on the posts list)
- Popover (tooltips, ellipsis menu, etc)

Ensure unit tests pass:

```
npm test
```
